ramanujantools/matrix.py

```diff
@@ -20,14 +21,65 @@ def __str__(self) -> str:
     def __eq__(self, other: Matrix) -> bool:
         return all(sp.simplify(cell) == 0 for cell in self - other)
 
-    def __call__(self, *args, **kwargs) -> Matrix:
+    def __hash__(self) -> int:
+        return hash(frozenset(self))
+
+    def __call__(self, substitutions: Dict) -> Matrix:
+        """
+        Substitutes symbols in the matrix, in a more math-like syntax.
         """
-        Substitutes variables in the matrix, in a more math-like syntax.
+        return self.subs(substitutions)
+
+    def subs(self, substitutions: Dict) -> Matrix:
+        """
+        Substitutes symbols in the matrix.
+
+        Calls the underlying sympy xreplace method, as subs is too slow and we don't need it's extra functionality
+        https://docs.sympy.org/latest/modules/core.html#sympy.core.basic.Basic.xreplace
 
-        Calls the underlying sympy `subs` method:
-        https://docs.sympy.org/latest/modules/core.html#sympy.core.basic.Basic.subs
+        In the case where all substitutions are numbers (and not symbols or expressions),
+        Uses numerical as an optimization
         """
-        return self.subs(*args, **kwargs)
+        if self._can_call_numerical_subs(substitutions):
+            return self.numerical_subs(substitutions)
+        return self.xreplace(substitutions)
+
+    def _can_call_numerical_subs(self, substitutions: Dict) -> bool:
+        """
+        Returns true iff the all substitutions are numerical and we can can call `numerical_subs` instead of `xreplace`.
+        """
+        return not (
+            len(substitutions) < len(self.free_symbols)
+            or any(isinstance(element, sp.Expr) for element in substitutions.values())
```

Comment:
What about `sp.Rational`, for example? Do we want floating point numerical evaluation? and if yes, we need to wrap it with parenthesis.

Reply:
Checked it, it works even for relatively complex expressions.